experimental/dds/tree2/src/test/feature-libraries/default-field-kinds/exhaustiveRebaserUtils.ts

```diff
@@ -0,0 +1,131 @@
+	) {
+		inputContext.push(current.mostRecentEdit.intention);
+	}
+	inputContext.reverse();
+	return inputContext;
+}
+
+export interface NamedChangeset<TChangeset> {
+	changeset: TaggedChange<TChangeset>;
+
+	/**
+	 * Intention for the changeset.
+	 */
+	intention: number;
+
+	/**
+	 * Friendly string describing the change.
+	 * This is typically used to name test cases.
+	 */
+	description: string;
+}
+
+/**
+ * Utility type for modeling valid series of edits to a field. Each node in the tree stores the contents
+ * of the field as well as information about the change applied to the parent state to get to this state.
+ * Because the tree is structured as an up-tree (stores parent pointers), the series of actions taken from
+ * some initial state to get to the state at a given node can be recovered by walking up the tree.
+ *
+ * This concept is similar to trees constructed in [state space search](https://en.wikipedia.org/wiki/State_space_search)
+ * or the tree implicit to the [minimax algorithm](https://en.wikipedia.org/wiki/Minimax).
+ *
+ * It's a useful data structure for testing rebasing because it allows more automatic test case generation, e.g.
+ * by exhaustively validating axioms against particular nodes in the tree up to a certain depth, or randomly selecting
+ * only certain nodes in the case of combinatorial explosion.
+ *
+ * This file largely contains helpers for selecting nodes of certain types in this tree.
+ * In order to use them with a particular field, that field must have an implementation of {@link ChildStateGenerator}.
+ */
+export interface FieldStateTree<TContent, TChangeset> {
+	content: TContent;
+	mostRecentEdit?: NamedChangeset<TChangeset>;
+	parent?: FieldStateTree<TContent, TChangeset>;
+}
+
+/**
+ * Given a particular state in a {@link FieldStateTree}, generates possible child states, i.e. states
+ * which can be reached by applying a single valid change.
+ */
+export type ChildStateGenerator<TContent, TChangeset> = (
+	state: FieldStateTree<TContent, TChangeset>,
+	tagFromIntention: (intention: number) => RevisionTag,
+	mintIntention: () => number,
+) => Iterable<FieldStateTree<TContent, TChangeset>>;
+
+function* walkFieldStateTree<TContent, TChangeset>(
+	initialState: FieldStateTree<TContent, TChangeset>,
+	generateChildStates: ChildStateGenerator<TContent, TChangeset>,
+	depth: number,
+	tagFromIntention: (intention: number) => RevisionTag,
+	mintIntention: () => number,
+): Iterable<FieldStateTree<TContent, TChangeset>> {
+	yield initialState;
+	if (depth > 0) {
+		for (const childState of generateChildStates(
+			initialState,
+			tagFromIntention,
+			mintIntention,
+		)) {
+			yield* walkFieldStateTree(
+				childState,
+				generateChildStates,
+				depth - 1,
+				tagFromIntention,
+				mintIntention,
+			);
+		}
+	}
+}
+
+function makeIntentionMinter(): () => number {
+	let intent = 0;
+	return () => intent++;
+}
+
+export function* generatePossibleSequenceOfEdits<TContent, TChangeset>(
```

Comment:
Would be nice to document the fact that it will generate them depth-first. Not a big deal though.

Reply:
since the sequences are fixed-length, depth-first/breadth-first doesn't make a difference here. I renamed the walk function though, as that does seem appropriate.